apps/src/templates/instructions/Instructions.jsx

```diff
@@ -74,26 +93,90 @@ class Instructions extends React.Component {
     }
   }
 
+  componentDidMount() {
+    let largestHeight = 0;
+
+    // Find the tallest of our dynamic instructions.
+    if (this.props.dynamicInstructions) {
+      Object.keys(this.props.dynamicInstructions).forEach(key => {
+        const height = $(
+          ReactDOM.findDOMNode(this.dynamicInstructionsRefs[key])
+        ).outerHeight(true);
+        if (height > largestHeight) {
+          largestHeight = height;
+        }
+      });
+    }
+
+    // Resize the parent div to be the height of the largest dynamic instruction.
+    if (this.props.setInstructionsRenderedHeight) {
+      this.props.setInstructionsRenderedHeight(
+        largestHeight + HEADER_HEIGHT + RESIZER_HEIGHT + 20 + 10
+      );
+    }
+  }
+
   render() {
+    const parentStyle = this.props.dynamicInstructions
+      ? styles.dynamicInstructions
+      : this.props.inTopPane
+      ? styles.inTopPane
+      : styles.notInTopPane;
+
+    const immersiveReaderText =
+      (this.props.dynamicInstructions &&
+        this.props.dynamicInstructions[this.props.dynamicInstructionsKey]) ||
+      this.props.longInstructions ||
+      this.props.shortInstructions;
+
     return (
-      <div
-        style={this.props.inTopPane ? styles.inTopPane : styles.notInTopPane}
-      >
+      <div style={parentStyle}>
         <ImmersiveReaderButton
           title={this.props.puzzleTitle || i18n.instructions()}
-          text={this.props.longInstructions || this.props.shortInstructions}
+          text={immersiveReaderText}
         />
-        {this.renderMainBody()}
-
-        {this.props.inputOutputTable && (
-          <InputOutputTable data={this.props.inputOutputTable} />
+        {this.props.dynamicInstructions && (
```

Comment:
It feels like maybe Dynamic Instructions for AI Lab should be there own component? Maybe AILabInstructions and just render that from TopInstructions instead of the Instructions component?

Reply:
Great call!  I've created a new `DynamicInstructions` component, and it feels much cleaner.